smoke_tests/tools/agent_runner.py

```diff
@@ -0,0 +1,234 @@
+
+        self._default_paths = default_paths
+
+    def _create_agent_files(self):
+        """
+        Create all essential files and directories and dynamically added files.
+        """
+        _make_or_clear_directory(self._agent_data_dir_path)
+
+        _make_or_clear_directory(self.agent_logs_dir_path)
+
+        for file_path in self._files.values():
+            self._create_file(file_path)
+
+        self.write_to_file(self._agent_config_path, json.dumps(self._agent_config))
+
+    def start(self):
+        # important to call this function before agent was started.
+        self._create_agent_files()
+
+        if self._installation_type == PACKAGE_INSTALL:
+            self._agent_process = subprocess.Popen(
+                ["/usr/sbin/scalyr-agent-2", "--no-fork", "--no-change-user", "start"],
+            )
+        else:
+            self._agent_process = subprocess.Popen(
+                "python -m scalyr_agent.agent_main --no-fork --no-change-user start",
+                shell=True,
+            )
+
+    def stop(self):
+        if self._agent_process.returncode is None:
+            self._agent_process.terminate()
+
+    @property
+    def _server_host(self):  # type: () -> six.text_type
+        return get_env("AGENT_HOST_NAME")
+
+    @property
+    def _agent_config(self):
+        # type: () -> Dict[six.text_type, Any]
+        """
+        Build and return agent configuration.
+        :return: dict with configuration.
+        """
+        return {
+            "verify_server_certificate": "false",
+            "server_attributes": {"serverHost": self._server_host},
+            "logs": list(self._log_files.values()),
+        }
+
+    @staticmethod
+    def _create_file(path, content=None):
+        # type: (Path, Optional[Any[six.text_type, six.binary_type]]) -> None
+        """
+        Add new file to runner's data directory.
+        :param path: path to new file, it is relative to runner's data directory path.
+        :param content: if set, write its data to file.
+        :return:
+        """
+
+        if path.exists():
+            os.remove(six.text_type(path))
+        if not path.parent.exists():
+            path.parent.mkdir(parents=True, exist_ok=True)
+        if not path.exists():
+            path.touch()
+
+        if content:
+            if isinstance(content, six.text_type):
+                path.write_text(content)
+            else:
+                path.write_bytes(content)
+
+    @staticmethod
+    def read_file_content(path):  # type: (Path) -> six.text_type
+        return path.read_text()
+
+    def write_to_file(self, path, data):
+        # type: (Path, Any[six.text_type, six.binary_type]) -> None
+        """
+        Write data to the file located in 'path'
+        :return:
+        """
+        data = six.ensure_text(data)
+
+        file_object = self._file_objects.get(six.text_type(path))
```

Comment:
Any particular reason why you are storing open file handles (for performance reasons or similar)?

To be on the safe side, we should make sure we always close those file handles (e.g. inside ``stop()`` or similar).

And if performance is not an issue, we could perhaps just use ``with`` context manager and open a file each time this method is called.

Reply:
EDIT: I see you use ``write_line`` in a loop here - https://github.com/scalyr/scalyr-agent-2/pull/401/files#diff-5eabc0cf126ed78d936ed2bafb25674eR212

So I assume that's the reason to avoid potentially man open system calls in that scenario?

Reply:
Yep, I thought it would be better to keep file handlers opened, but now I think 1000 opens and writes should be fine too. Also, this can be additional stress test for agent log readers. What do you think?